worlds/yoshisisland/Rules.py

```diff
@@ -0,0 +1,789 @@
+    set_rule(multiworld.get_location("Tap-Tap The Red Nose's Fort: Stars", player), lambda state: logic.has_midring(state) and state.has_all({'Spring Ball', 'Egg Plant', 'Key'}, player))
+
+    set_rule(multiworld.get_location("The Very Loooooong Cave: Red Coins", player), lambda state: state.has_all({'Chomp Rock'}, player) and (state.has('Egg Capacity Upgrade', player, 2) or logic.combat_item(state)))
+    set_rule(multiworld.get_location("The Very Loooooong Cave: Flowers", player), lambda state: state.has_all({'Chomp Rock'}, player) and (state.has('Egg Capacity Upgrade', player, 2) or logic.combat_item(state)))
+    set_rule(multiworld.get_location("The Very Loooooong Cave: Stars", player), lambda state: state.has_all({'Chomp Rock'}, player) and (state.has('Egg Capacity Upgrade', player, 2) or logic.combat_item(state)) and logic.has_midring(state))
+    set_rule(multiworld.get_location("The Very Loooooong Cave: Level Clear", player), lambda state: True)
+
+    set_rule(multiworld.get_location("The Deep, Underground Maze: Red Coins", player), lambda state: state.has_all({'Chomp Rock', 'Key', 'Large Spring Ball'}, player))
+    set_rule(multiworld.get_location("The Deep, Underground Maze: Flowers", player), lambda state: state.has_all({'Chomp Rock', 'Key', 'Large Spring Ball'}, player))
+    set_rule(multiworld.get_location("The Deep, Underground Maze: Stars", player), lambda state: state.has_all({'Chomp Rock', 'Tulip', 'Key'}, player) or (logic.has_midring(state) and state.has_all({'Key', 'Chomp Rock', 'Large Spring Ball'}, player)))
+    set_rule(multiworld.get_location("The Deep, Underground Maze: Level Clear", player), lambda state: state.has_all({'Key', 'Large Spring Ball', 'Dashed Platform'}, player) and (logic.combat_item(state) or state.has('Chomp Rock', player)))
+
+    set_rule(multiworld.get_location("KEEP MOVING!!!!: Red Coins", player), lambda state: True)
+    set_rule(multiworld.get_location("KEEP MOVING!!!!: Flowers", player), lambda state: True)
+    set_rule(multiworld.get_location("KEEP MOVING!!!!: Stars", player), lambda state: logic.has_midring(state))
+    set_rule(multiworld.get_location("KEEP MOVING!!!!: Level Clear", player), lambda state: True)
+
+    set_rule(multiworld.get_location("King Bowser's Castle: Red Coins", player), lambda state: state.has_all({'Helicopter Morph', 'Egg Plant'}, player) and logic._68CollectibleRoute(state))
+    set_rule(multiworld.get_location("King Bowser's Castle: Flowers", player), lambda state: state.has_all({'Helicopter Morph', 'Egg Plant'}, player) and logic._68CollectibleRoute(state))
+    set_rule(multiworld.get_location("King Bowser's Castle: Stars", player), lambda state: state.has_all({'Helicopter Morph', 'Egg Plant'}, player) and logic._68Route(state))
+
+    if world.options.extras_enabled == 1:

```

Comment:
could probably make a method out of this block

Reply:
Could you elaborate?

Reply:
I'm not sure how I can be more specific other than writing the whole code

```py
# [...]
def set_easy_rules(world, multiworld, player: int, boss_order, luigi_pieces):
    logic = YoshiLogic(multiworld, player, boss_order, luigi_pieces, world)
    
    # [...]
    
        set_rule(multiworld.get_location("King Bowser's Castle: Stars", player), lambda state: state.has_all({'Helicopter Morph', 'Egg Plant'}, player) and logic._68Route(state))
        
        set_easy_extra_rules()
        
def set_easy_extra_rules(world, multiworld, player, ):
        if world.options.extras_enabled != 1:
            return
        set_rule(multiworld.get_location("Poochy Ain't Stupid: Red Coins", player), lambda state: state.has_all({'Poochy'}, player))
        set_rule(multiworld.get_location("Poochy Ain't Stupid: Flowers", player), lambda state: state.has_all({'Poochy'}, player))
        set_rule(multiworld.get_location("Poochy Ain't Stupid: Stars", player), lambda state: state.has('Poochy', player))
        set_rule(multiworld.get_location("Poochy Ain't Stupid: Level Clear", player), lambda state: state.has('Poochy', player))

        set_rule(multiworld.get_location("Hit That Switch!!: Red Coins", player), lambda state: state.has_all({'Large Spring Ball', '! Switch'}, player))
        set_rule(multiworld.get_location("Hit That Switch!!: Flowers", player), lambda state: state.has_all({'Large Spring Ball', '! Switch'}, player))
        set_rule(multiworld.get_location("Hit That Switch!!: Stars", player), lambda state: True)
        set_rule(multiworld.get_location("Hit That Switch!!: Level Clear", player), lambda state: state.has_all({'Large Spring Ball', '! Switch'}, player))

        set_rule(multiworld.get_location("More Monkey Madness: Red Coins", player), lambda state: True)
        set_rule(multiworld.get_location("More Monkey Madness: Flowers", player), lambda state: True)
        set_rule(multiworld.get_location("More Monkey Madness: Stars", player), lambda state: True)
        set_rule(multiworld.get_location("More Monkey Madness: Level Clear", player), lambda state: True)

        set_rule(multiworld.get_location("The Impossible? Maze: Red Coins", player), lambda state: state.has_all({'Spring Ball', 'Large Spring Ball', 'Mole Tank Morph', 'Helicopter Morph', 'Flashing Eggs'}, player))
        set_rule(multiworld.get_location("The Impossible? Maze: Flowers", player), lambda state: state.has_all({'Spring Ball', 'Large Spring Ball', 'Mole Tank Morph', 'Helicopter Morph'}, player))
        set_rule(multiworld.get_location("The Impossible? Maze: Stars", player), lambda state: state.has_all({'Spring Ball', 'Large Spring Ball', 'Mole Tank Morph'}, player))
        set_rule(multiworld.get_location("The Impossible? Maze: Level Clear", player), lambda state: state.has_all({'Spring Ball', 'Large Spring Ball', 'Mole Tank Morph', 'Helicopter Morph'}, player))

        set_rule(multiworld.get_location("Kamek's Revenge: Red Coins", player), lambda state: state.has_all({'Key', 'Skis', 'Helicopter Morph', '! Switch'}, player) and logic.has_midring(state))
        set_rule(multiworld.get_location("Kamek's Revenge: Flowers", player), lambda state: state.has_all({'Key', 'Skis', 'Helicopter Morph', '! Switch'}, player) and logic.has_midring(state))
        set_rule(multiworld.get_location("Kamek's Revenge: Stars", player), lambda state: state.has_all({'! Switch'}, player) and logic.has_midring(state))
        set_rule(multiworld.get_location("Kamek's Revenge: Level Clear", player), lambda state: state.has_all({'Key', 'Skis', '! Switch', 'Helicopter Morph'}, player))

        set_rule(multiworld.get_location("Castles - Masterpiece Set: Red Coins", player), lambda state: (state.has('Egg Capacity Upgrade', player, 2) or logic.combat_item(state)) and state.has(('Large Spring Ball'), player))
        set_rule(multiworld.get_location("Castles - Masterpiece Set: Flowers", player), lambda state: (state.has('Egg Capacity Upgrade', player, 2) or logic.combat_item(state)) and state.has(('Large Spring Ball'), player))
        set_rule(multiworld.get_location("Castles - Masterpiece Set: Stars", player), lambda state: logic.has_midring(state) and state.has(('Large Spring Ball'), player))
        set_rule(multiworld.get_location("Castles - Masterpiece Set: Level Clear", player), lambda state: (state.has('Egg Capacity Upgrade', player, 2) or logic.combat_item(state)) and state.has(('Large Spring Ball'), player))

def set_normal_rules(world, multiworld, player: int, boss_order, luigi_pieces):
    # [...]
```

Reply:
Ah, I understand now. Is this more efficient than having it part of the regular easy rules method

Reply:
Generally, splitting things into method based on behavioral branches is considered good practice.
In this example, one big appeal is the ability to invert the if, which reduces nesting and therefore makes it easier for the person reading the code, as they have fewer conditions to keep in mind.
Essentially, these rules all have something in common, which can be expressed in a single method call with a good name, and allow anyone reading the calling code to "skip over" the block if they aren't interested in it, or to go straight to it if it's what they are looking for